### PR TITLE
Work branch

### PR DESCRIPTION
**Pull request by Jojosito**
- Enhance d4pi
- rename coordinates -> amplitude
- automatic determination of optimal lag
- program to retrieve generated fit fractions
- programs to compare scaling of Zemach and Helicity amplitudes

### CODE REVIEW

File: examples/bat_gen/bat_fit.cxx [hunk 1]
```diff
@@ -73,10 +73,12 @@ bat_fit::bat_fit(std::string name, std::unique_ptr<yap::Model> M, const std::vec
     //                         [](int n, const yap::IntegralMap::value_type& v)
     //                         {return n + v.second.decayTrees().size();});
     // if (N > 1) {
+    unsigned i(0); // running index to avoid name collisions
     for (const auto& b2_dtvi : Integral_.integrals())
         for (const auto& dt : b2_dtvi.second.decayTrees()) {
             DecayTrees_.push_back(dt);
-            AddObservable("fit_frac(" + to_string(*dt->freeAmplitude()->decayChannel()) + " M = " + yap::spin_to_string(dt->freeAmplitude()->twoM()) + ")", 0, 1.1);
+            AddObservable("fit_frac(" + to_string(*dt->freeAmplitude()) + ")" + std::to_string(i++), 0, 1.1);
```
Comment (greenwald) [inline examples/bat_gen/bat_fit.cxx]: this should made to give a sensible line.

File: examples/bat_gen/bat_fit.cxx [hunk 2]
```diff
@@ -272,8 +293,10 @@ size_t bat_fit::findFreeAmplitude(std::shared_ptr<yap::FreeAmplitude> A) const
 {
     auto it = std::find(FreeAmplitudes_.begin(), FreeAmplitudes_.end(), A);
 
-    if (it == FreeAmplitudes_.end())
+    if (it == FreeAmplitudes_.end()) {
+        LOG(INFO) << "did not find amplitude " << A;
```
Comment (greenwald) [inline examples/bat_gen/bat_fit.cxx]: Why the LOG output?

File: examples/bat_gen/bat_fit.h [hunk 3]
```diff
@@ -75,6 +75,9 @@ class bat_fit : public bat_yap_base
     yap::DataPartitionVector& integralPartitions()
     { return IntegralPartitions_; }
 
+    yap::ModelIntegral& modelIntegral()
```
Comment (greenwald) [inline examples/bat_gen/bat_fit.h]: `integral()`

File: examples/bat_gen/bat_fit.h [hunk 4]
```diff
@@ -99,6 +102,10 @@ class bat_fit : public bat_yap_base
     /// find the position in the parameter list of the first element of a free amplitude
     size_t findFreeAmplitude(std::shared_ptr<yap::FreeAmplitude> A) const;
 
+    /// \return free amplitudes
+    const yap::FreeAmplitudeVector freeAmplitudes() const
```
Comment (greenwald) [inline examples/bat_gen/bat_fit.h]: `&`

File: examples/bat_gen/models/d4pi.h [hunk 5]
```diff
@@ -69,17 +83,26 @@ inline std::unique_ptr<Model> d4pi(std::unique_ptr<yap::Model> M)
     a_1->addChannel(sigma, piPlus);
 
     // a_1 -> rho pi
-    *free_amplitude(*a_1, to(rho), l_equals(0)) = 1; // S wave
-    free_amplitude(*a_1, to(rho), l_equals(0))->variableStatus() = VariableStatus::fixed;
-    *free_amplitude(*a_1, to(rho), l_equals(1)) = 0.; // P wave
-    *free_amplitude(*a_1, to(rho), l_equals(2)) = std::polar(0.241, rad(82.)); // D wave
+    for (auto& f : free_amplitudes(*a_1, to(rho), l_equals(0))) {
```
Comment (greenwald) [inline examples/bat_gen/models/d4pi.h]: `free_amplitude(...)`

File: examples/bat_gen/models/d4pi.h [hunk 6]
```diff
@@ -91,7 +114,7 @@ inline std::unique_ptr<Model> d4pi(std::unique_ptr<yap::Model> M)
     f_2->addChannel(piPlus, piMinus); 
     
     // pi+ pi- flat
-    auto pipiFlat = F.decayingParticle(F.pdgCode("f_0"), radialSize); // just need any spin0 particle
+    auto pipiFlat = F.decayingParticle(F.pdgCode("pi0"), radialSize); // just need any spin0 particle
```
Comment (greenwald) [inline examples/bat_gen/models/d4pi.h]: `DecayingParticle::create("pipiflat", QuantumNumbers(0, 0),  radialSize);`

File: examples/bat_gen/models/d4pi.h [hunk 7]
```diff
@@ -121,10 +144,123 @@ inline std::unique_ptr<Model> d4pi(std::unique_ptr<yap::Model> M)
...
+    auto D     = std::static_pointer_cast<DecayingParticle>(particle(*m.model(), is_named("D0")));
+    auto rho   = std::static_pointer_cast<Resonance>(particle(*m.model(), is_named("rho0")));
+    auto sigma = std::static_pointer_cast<Resonance>(particle(*m.model(), is_named("f_0(500)")));
+    auto a_1   = std::static_pointer_cast<Resonance>(particle(*m.model(), is_named("a_1+")));
+    auto f_0   = std::static_pointer_cast<Resonance>(particle(*m.model(), is_named("f_0")));
+    auto f_2   = std::static_pointer_cast<Resonance>(particle(*m.model(), is_named("f_2")));
+
+
+    auto fixed_amps = free_amplitudes(*a_1, to(rho), l_equals(0));
+    for (auto fa : fixed_amps) {
```
Comment (greenwald) [inline examples/bat_gen/models/d4pi.h]: `const auto&`

File: examples/bat_gen/runBatFit.cxx [hunk 8]
```diff
@@ -36,8 +37,10 @@ int main()
     yap::plainLogs(el::Level::Info);
 
     // open file
-    std::string model_name = "D3PI";
-    // std::string model_name = "DKKPI";
+    // std::string model_name = "D3PI";
+    //std::string model_name = "DKKPI";
+    std::string model_name = "D4PI";
```
Comment (greenwald) [inline examples/bat_gen/runBatFit.cxx]: add 4pi as a commented out option;
or better yet: add switches to determine. `if (true) { ... } else { ... }`

File: examples/programs/CMakeLists.txt [hunk 9]
```diff
@@ -1,7 +1,13 @@
-include_directories(${YAP_SOURCE_DIR}/include)
+include_directories(${YAP_SOURCE_DIR}/include ${ROOT_INCLUDE_DIR})
```
Comment (greenwald) [inline examples/programs/CMakeLists.txt]: let's not add ROOT back in unless REALLY needed.

File: examples/bat_gen/models/d4pi.h (hunk 5 above)
Comment (Jojosito) [inline examples/bat_gen/models/d4pi.h]: Container size not 1 (3) from lone_elt

File: examples/bat_gen/bat_fit.cxx (hunk 1 above)
Comment (Jojosito) [inline examples/bat_gen/bat_fit.cxx]: When there are two decay trees which only differ by some daughters' spin projections, e.g.
`fit_frac(rho0 rho0, M = 0, 0 -> 1 + 1 with L = 1 and S = 1;     d[0] --> pi+ pi-, M = -1, 1 -> 0 + 0 with L = 1 and S = 0 [fixed];     d[1] --> pi+ pi-, M = -1, 1 -> 0 + 0 with L = 1 and S = 0 [fixed])`
and
`fit_frac(rho0 rho0, M = 0, 0 -> 1 + 1 with L = 1 and S = 1;     d[0] --> pi+ pi-, M = 1, 1 -> 0 + 0 with L = 1 and S = 0 [fixed];     d[1] --> pi+ pi-, M = 1, 1 -> 0 + 0 with L = 1 and S = 0 [fixed])`,
the minus signs will be omitted in BAT's safe name, and there is still a name collision:
`Error   : BCVariableSet::Add : Variable with safe name fit_fracrho0rho0M0011withL1andS1d0pipiM1100withL1andS0fixedd1pipiM1100withL1andS0fixed%s exists already.`

Comment (Jojosito) [inline examples/bat_gen/bat_fit.cxx]: Replacing the - with m solves the issue, but its a bit hackish ...

File: examples/bat_gen/models/d4pi.h [hunk 10]
```diff
@@ -69,17 +83,30 @@ inline std::unique_ptr<Model> d4pi(std::unique_ptr<yap::Model> M)
...
+    }
+    /**free_amplitude(*a_1, to(rho), l_equals(0)) = 1;
+    *free_amplitude(*a_1, to(rho), l_equals(1)) = 0.;
+    free_amplitude(*a_1, to(rho), l_equals(1))->variableStatus() = VariableStatus::fixed;
+    *free_amplitude(*a_1, to(rho), l_equals(2)) = std::polar(0.241, rad(82.));*/
 
     // a_1 -> sigma pi 
-    *free_amplitude(*a_1, to(sigma)) = std::polar(0.439, rad(193.));
-    
+    for (auto& f : free_amplitudes(*a_1, to(sigma)))
```
Comment (greenwald) [inline examples/bat_gen/models/d4pi.h]: `free_amplitude`

File: examples/bat_gen/models/d4pi.h [hunk 11]
```diff
@@ -121,10 +148,123 @@ inline std::unique_ptr<Model> d4pi(std::unique_ptr<yap::Model> M)
     for (unsigned l = 0; l < 3; ++l) {
         auto freeAmp = free_amplitude(*M, to(rho, rho), l_equals(l));
         LOG(INFO) << to_string(*freeAmp);
-        *freeAmp = c[l];
+        *freeAmp = std::complex<double>(c[l]);
```
Comment (greenwald) [inline examples/bat_gen/models/d4pi.h]: `static_cast<std::complex<double> >(c[1])`

**Pull request closed without merging**